src/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/RequestExecutionQueue.cs

```diff
@@ -245,10 +246,27 @@ private async Task ProcessQueueAsync()
                     // notifications have been completed by the time we attempt to determine the language, so we have the up to date map of URI to language.
                     // Since didOpen notifications are marked as mutating, the queue will not advance to the next request until the server has finished processing
                     // the didOpen, ensuring that this line will only run once all prior didOpens have completed.
-                    var language = _languageServer.GetLanguageForRequest(work.MethodName, work.SerializedRequest);
+                    var didGetLanguage = _languageServer.TryGetLanguageForRequest(work.MethodName, work.SerializedRequest, out var language);
 
                     // Now that we know the actual language, we can deserialize the request and start creating the request context.
-                    var (metadata, handler, methodInfo) = GetHandlerForRequest(work, language);
+                    var (metadata, handler, methodInfo) = GetHandlerForRequest(work, language ?? LanguageServerConstants.DefaultLanguageName);
+
+                    // We had an issue determining the language.  Generally this is very rare and only occurs
+                    // if there is a mis-behaving client that sends us requests for files where we haven't saved the languageId.
+                    // We should only crash if this was a mutating method, otherwise we should just fail the single request. 
+                    if (!didGetLanguage)
+                    {
+                        var exception = new InvalidOperationException($"Failed to get language for {work.MethodName}");
+                        if (handler.MutatesSolutionState)
+                        {
+                            throw exception;
+                        }
+                        else
+                        {
+                            work.FailRequest(exception);
```

Comment:
slightly odd.  would prefer a message here.  the impl of it can make an exception if you want.  then the `new Invalid...` also just moves into the Mutates case.